modules/git/git_api.cpp

```diff
@@ -0,0 +1,296 @@
+#include "git_api.h"
+
+#include "core/os/file_access.h"
+#include "core/os/os.h"
+#include "editor/plugins/version_control_editor_plugin.h"
+#include "editor/editor_file_system.h"
+
+EditorVCSInterfaceGit *EditorVCSInterfaceGit::singleton = nullptr;
+
+void EditorVCSInterfaceGit::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("_initialize", "project_root_path"), &EditorVCSInterfaceGit::_initialize);
+	ClassDB::bind_method(D_METHOD("_is_vcs_initialized"), &EditorVCSInterfaceGit::_is_vcs_initialized);
+	ClassDB::bind_method(D_METHOD("_get_vcs_name"), &EditorVCSInterfaceGit::_get_vcs_name);
+	ClassDB::bind_method(D_METHOD("_shut_down"), &EditorVCSInterfaceGit::_shut_down);
+	ClassDB::bind_method(D_METHOD("_get_project_name"), &EditorVCSInterfaceGit::_get_project_name);
+	ClassDB::bind_method(D_METHOD("_get_modified_files_data"), &EditorVCSInterfaceGit::_get_modified_files_data);
+	ClassDB::bind_method(D_METHOD("_commit", "msg"), &EditorVCSInterfaceGit::_commit);
+	ClassDB::bind_method(D_METHOD("_get_file_diff", "file_path"), &EditorVCSInterfaceGit::_get_file_diff);
+	ClassDB::bind_method(D_METHOD("_stage_file", "file_path"), &EditorVCSInterfaceGit::_stage_file);
+	ClassDB::bind_method(D_METHOD("_unstage_file", "file_path"), &EditorVCSInterfaceGit::_unstage_file);
+
+	ClassDB::bind_method(D_METHOD("_setup"), &EditorVCSInterfaceGit::_setup);
+}
+
+void EditorVCSInterfaceGit::_setup() {
+	// EditorVCSInterfaceGit singleton must NOT be assigned until `register_editor` is called.
+	// otherwise no docks will appear.
+	VersionControlEditorPlugin::get_singleton()->register_editor();
```

Comment:
It seems what this function does is that it tries to replace the object instance which the editor uses to do all the VCS operations.

However, can this actually leak memory if you don't destroy the original `EditorVCSInterface` instance that was present there before assigning it to an instance of `EditorVCSInterfaceGit` from your module?

Reply:
I think this could happen. But initialization of `EditorVCSInterfaceGit` starts before any other plugin, and as far as I know, no VCS is initialized at editor startup even when using GDNative, referring to godotengine/godot#49309.

Reply:
I see, still, I don't think it would hurt to see if the singleton is existing, if it does then delete it and assign it to an object of your own. The editor takes care of when to destroy the plugin anyway.